cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,279 @@
+    }
+
+    quantile_index idx(size, quantile);
+
+    T a;
+    T b;
+    TResult value;
+
+    switch (interpolation) {
+    case interpolation::LINEAR:
+        a = get_array_value(begin, sortmap[idx.lower_bound]);
+        b = get_array_value(begin, sortmap[idx.upper_bound]);
+        cudf::interpolate::linear<TResult>(value, a, b, idx.fraction);
+        break;
+
+    case interpolation::MIDPOINT:
+        a = get_array_value(begin, sortmap[idx.lower_bound]);
+        b = get_array_value(begin, sortmap[idx.upper_bound]);
+        cudf::interpolate::midpoint<TResult>(value, a, b);
+        break;
+
+    case interpolation::LOWER:
+        a = get_array_value(begin, sortmap[idx.lower_bound]);
+        value = static_cast<TResult>(a);
+        break;
+
+    case interpolation::HIGHER:
+        a = get_array_value(begin, sortmap[idx.upper_bound]);
+        value = static_cast<TResult>(a);
+        break;
+
+    case interpolation::NEAREST:
+        a = get_array_value(begin, sortmap[idx.nearest]);
+        value = static_cast<TResult>(a);
+        break;
+
+    default:
+        throw new cudf::logic_error("not implemented");
+    }
+
+    return std::make_unique<numeric_scalar<TResult>>(value);
+}
+
+template<typename T, typename TResult>
+std::unique_ptr<scalar>
+trampoline(column_view const& in,
+           double quantile,
+           bool is_sorted,
+           order order,
+           null_order null_order,
+           interpolation interpolation,
+           rmm::mr::device_memory_resource *mr,
+           cudaStream_t stream)
+{
+    if (in.size() == 1) {
+        auto result_value = get_array_value(in.begin<T>(), 0);
+        return std::make_unique<numeric_scalar<TResult>>(static_cast<TResult>(result_value));    }
+
+    if (not is_sorted) {
+
+        // using TScalar = cudf::experimental::scalar_type_t<T>;
+        // data_type type{type_to_id<T>()};
+
+        if (quantile >= 1.0)
+        {
+            // T const * value_p = thrust::max_element(rmm::exec_policy(stream)->on(stream), data, data + size);
+            // T value = cudf::detail::get_array_value(value_p, 0);
+            // auto result = make_numeric_scalar(type);
+            // auto result_sc = static_cast<TScalar *>(result.get());
+            // result_sc->set_value(value);
+            // return result;
+        }
+
+        if (quantile <= 0.0)
+        {
+            // T const * value_p = thrust::min_element(rmm::exec_policy(stream)->on(stream), data, data + size);
+            // T value = cudf::detail::get_array_value(value_p, 0);
+            // auto result = make_numeric_scalar(type);
+            // auto result_sc = static_cast<TScalar *>(result.get());
+            // result_sc->set_value(value);
+            // return result;
+        }
+
+        table_view unsorted {{ in }};
+        auto sorted_idx = sorted_order(unsorted, { order }, { null_order });
+        auto sorted = gather(unsorted, sorted_idx->view());
```

Comment:
Is it necessary to actually materialize the sorted result? That can be quite expensive. Consider just using the sorted_order to do indirect lookups into the input.

Reply:
No, it's not necessary, and I've been looking for a way to get around it. I'd rather use `sorted_idx` and `begin` together in a permutation iterator, but I'm still wrapping my brain around host/device interaction. My intuition says thrust can handle this, but I haven't figured it out yet.